core/modules/shrscan/Director.h

```diff
@@ -0,0 +1,315 @@
+
+class Director
+{
+public:
+
+//-----------------------------------------------------------------------------
+//! @brief Add a query to the shared scan queue.
+//!
+//! @param  query - Pointer to a Query object that describes the query.
+//-----------------------------------------------------------------------------
+
+virtual void  addQuery(std::shared_ptr<Query> const& query) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Cancel a shared scan query.
+//!
+//! When a query is cancelled, the Query object's endQuery() method is called.
+//! If the query is executing its runQuery() method, cancellation occurs upon
+//! return.
+//!
+//! @param  query   - Pointer to a Query object that describes the query.
+//! @param  wait    - When true, does not return until the query has been
+//!                   actually cancelled (i.e. endQuery() method called).
+//!
+//! @return false: The query was not found.
+//! @return true:  The query was cancelled or was scheduled to be cancelled.
+//-----------------------------------------------------------------------------
+
+virtual bool  cancelQuery(Query const* query, bool wait=false) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Cancel all shared scan queries.
+//!
+//! @param  wait    - When true, does not return until all queries have been
+//!                   actually cancelled (i.e. endQuery() method called).
+//!
+//! This method effectively calls cancelQuery() on each query.
+//-----------------------------------------------------------------------------
+
+virtual void  cancelAll(bool wait=false) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Obtain statistics about this shared scan director.
+//!
+//! @param  stats   - Where information is to be returned (see struct below).
+//!
+//! @return The number of queries actually running.
+//-----------------------------------------------------------------------------
+
+struct Statistics {
+    unsigned long long bytesLockMax;//< Maximum number of bytes to lock
+    unsigned long long bytesLocked; //< Actual  number of bytes locked in memory
+    unsigned long long msRunTotal;  //< Total milliseconds spent running
+    unsigned long long msIdleTotal; //< Total milliseconds spent idling
+    unsigned int       numChunks;   //< The number of chunks being handled
+    unsigned int       maxCanRun;   //< Maximum number of parallel queries
+    unsigned int       numQueued;   //< Number waiting to run
+    unsigned int       numRunning;  //< Number actually running
+    unsigned int       numSuspend;  //< Number suspended
+    unsigned int       numErrors;   //< Number of queries that had errors
+    unsigned int       numCancelled;//< Number of queries that were cancelled
+};
+
+virtual void  getStatistics(Statistics &stats) = 0;
```

Comment:
OK, you've convinced me :-)

On Thu, 19 Nov 2015, Andy wrote:

>> +
>> +struct Statistics {
>> +    unsigned long long bytesLockMax;//< Maximum number of bytes to lock
>> +    unsigned long long bytesLocked; //< Actual  number of bytes locked in memory
>> +    unsigned long long msRunTotal;  //< Total milliseconds spent running
>> +    unsigned long long msIdleTotal; //< Total milliseconds spent idling
>> +    unsigned int       numChunks;   //< The number of chunks being handled
>> +    unsigned int       maxCanRun;   //< Maximum number of parallel queries
>> +    unsigned int       numQueued;   //< Number waiting to run
>> +    unsigned int       numRunning;  //< Number actually running
>> +    unsigned int       numSuspend;  //< Number suspended
>> +    unsigned int       numErrors;   //< Number of queries that had errors
>> +    unsigned int       numCancelled;//< Number of queries that were cancelled
>> +};
>> +
>> +virtual void  getStatistics(Statistics &stats) = 0;
>
> OK, one more reason why return-by-value is better and I stop trying to convince you :smile:
>
> In this code:
>
>    Statistics stat;         // 1
>    getStatistics(stat);     // 2
>
> There is _always_ a construction happening (in 1) and copy (in 2), there is no way to avoid this two-step process (in you particular case construction does not do anything but in general this is not always true, depends entirely on the class).
>
> In this code:
>
>    Statistics stat = getStatistics(stat);
>
> if method is written optimally and all optimizations are in place there is only one constructor call happening, no additional copies (it can be a copy constructor).
>
> If you care about efficiency then you should think about avoiding unnecessary initialization too.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lsst/qserv/pull/156/files#r45433851